File: 0314-binary-tree-vertical-order-traversal/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    
    def verticalOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        
        nodeVertices = dict()
        
        def preOrderTraversal(node, vertLevel, depth):
            if node == None:
                return
            
            if node == root:
                nodeVertices[0] = [(root.val, depth)]
            elif node:
                if vertLevel not in nodeVertices:
                    nodeVertices[vertLevel] = []
                nodeVertices[vertLevel].append((node.val, depth))
            
            preOrderTraversal(node.left, vertLevel - 1, depth+1)
            preOrderTraversal(node.right, vertLevel + 1, depth+1)
            return 
        
        minVertical = 0
        preOrderTraversal(root, minVertical,0)
        
        
        result = []
        vertLev = min(nodeVertices.keys()) if len(nodeVertices) else -1      
    
        while vertLev in nodeVertices:
            vertices = [item[0] for item in sorted(nodeVertices[vertLev], key=lambda x: x[1])]
            
            result.append(vertices)
            vertLev += 1
        return result
                

File: 0314-binary-tree-vertical-order-traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalOrder(TreeNode* root) 
    {
        if(root == nullptr) return {};
        
        
        map<int, vector<int>> nodePos;
        
        queue<pair<TreeNode*, int>> Q;
        Q.push({root, 0});
        
        int col;
        while(!Q.empty())
        {
            pair<TreeNode*, int> item = Q.front();
            Q.pop();
            
            col        = item.second;
            TreeNode* node = item.first;
            
            nodePos[col].push_back(node->val);
            
            if(node->left) Q.push({node->left, col-1});
            
            if(node->right) Q.push({node->right, col+1});
        }
        
        
        vector<vector<int>> ans;
        
        for(auto itr = nodePos.begin(); itr != nodePos.end(); itr++)
        {
            ans.push_back(itr->second);
        }
        
        return ans;
        
        
    }
};
